src/main/java/org/ow2/proactive/connector/iaas/cloud/provider/jclouds/aws/AWSEC2JCloudsProvider.java

```diff
@@ -222,4 +284,83 @@ public void removeInstancePublicIp(Infrastructure infrastructure, String instanc
+            return null;
+        }
+    }
+
+    private KeyPairApi getKeyPairApi(Infrastructure infrastructure) {
+        ComputeService computeService = getComputeServiceFromInfastructure(infrastructure);
+        EC2Api ec2Api = computeService.getContext().unwrapApi(EC2Api.class);
+        if (ec2Api.getKeyPairApi().isPresent()) {
+            return ec2Api.getKeyPairApi().get();
+        } else {
+            throw new UnsupportedOperationException("Cannot retrieve AWS key pair API, which enables key pair creation");
+        }
+    }
+
+    private String getRegionFromImage(Instance instance) {
+        String image = instance.getImage();
+        return image.split(INSTANCE_ID_REGION_SEPARATOR)[0];
+    }
+
+    @Override
+    public RunScriptOptions getDefaultRunScriptOptions(String instanceId, Infrastructure infrastructure,
+            String instanceTag) {
+        ComputeService computeService = getComputeServiceFromInfastructure(infrastructure);
+        NodeMetadata node = retrieveNodeMetadataFrom(instanceId, instanceTag, computeService);
+        String instanceRegion = getRegionFromNode(computeService, node);
+
+        // the region returned here contains a subdivision of the region, for
+        // example eu-west-1c for the region eu-west-1, so we need to remove
+        // the subdivision to have the exact region name of the key
+        String keyPairRegion = instanceRegion.substring(0, instanceRegion.length() - 1);
+        logger.debug("Using region " + keyPairRegion + " for instance id " + instanceId);
+
+        // we use the key pair that was created in the AWS region of the instance
+        return Optional.ofNullable(generatedKeyPairsPerAwsRegion.get(keyPairRegion))
+                       // currently, we cannot login as root user on ec2 instances anymore
+                       // this is why we absolutely need to pass an authentication key
+                       .map(keypair -> RunScriptOptions.Builder.runAsRoot(false)
+                                                               .overrideLoginUser(getVmUserLogin())
+                                                               .overrideLoginPrivateKey(keypair.getValue()))
+                       // fall back, last helpless chance to run the script
+                       .orElse(RunScriptOptions.NONE);
+    }
+
+    private NodeMetadata retrieveNodeMetadataFrom(String instanceId, String instanceTag,
+            ComputeService computeService) {
+        NodeMetadata node = null;
+        if (instanceId != null) {
+            node = computeService.getNodeMetadata(instanceId);
+        } else {
+            if (instanceTag != null) {
+                // here we attempt to retrieve at least one node that has the given tag
+                Set<? extends NodeMetadata> allNodes = computeService.listNodesDetailsMatching(all());
+                for (NodeMetadata checkedNode : allNodes) {
+                    if (checkedNode.getGroup().equals(instanceTag)) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AS134)